zerver/actions/message_edit.py

```diff
@@ -147,20 +147,31 @@ def maybe_send_resolve_topic_notifications(
     old_topic_name: str,
     new_topic_name: str,
     changed_messages: QuerySet[Message],
+    pre_truncation_new_topic_name: str,
 ) -> Tuple[Optional[int], bool]:
     """Returns resolved_topic_message_id if resolve topic notifications were in fact sent."""
     # Note that topics will have already been stripped in check_update_message.
     #
     # This logic is designed to treat removing a weird "✔ ✔✔ "
     # prefix as unresolving the topic.
-    topic_resolved: bool = new_topic_name.startswith(
-        RESOLVED_TOPIC_PREFIX
-    ) and not old_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
-    topic_unresolved: bool = old_topic_name.startswith(
-        RESOLVED_TOPIC_PREFIX
-    ) and not new_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
+    resolved_prefix_len = len(RESOLVED_TOPIC_PREFIX)
+    topic_resolved: bool = (
+        new_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
+        and not old_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
+        and pre_truncation_new_topic_name[resolved_prefix_len:] == old_topic_name
+    )
+    topic_unresolved: bool = (
+        old_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
+        and not new_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
+        and old_topic_name.lstrip(RESOLVED_TOPIC_PREFIX) == new_topic_name
+    )
 
     if not topic_resolved and not topic_unresolved:
+        # This makes the assumption that a stream can't have two topics
+        # of the same name with one unresolved and another resolved,
+        # because if so the topic will be considered resolved or unresolved
+        # when a message is moved between the two topics.
```

Comment:
So this assumption is not true; it's very much possible to send a message into the non-resolved version of a resolved topic. We will likely change this at some point by changing the encoding of resolved topics to be a separate boolean from the topic name; while we originally thought this included-in-name design was ideal, experience has made us feel like it was the wrong approach.

Reply:
This is based on [CZO Discussion ](https://chat.zulip.org/#narrow/stream/9-issues/topic/.F0.9F.93.82.20bug.20when.20moving.20a.20message.20out.20of.20resolved.20topic/near/1799165)

Reply:
OK, I think the logic is correct as designed there, though; the implementation here doesn't try to do anything special, but that's the most reasonable thing to do with the current data model.